src/System.Security.Cryptography.Encoding/tests/Asn1/Reader/PeekTests.cs

```diff
@@ -133,5 +134,80 @@ public static void PeekContentSpan_Corrupt_Throws()
                     reader.PeekContentBytes();
                 });
         }
+
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public static void PeekContentSpan_ExtremelyNested(bool fullArray)
+        {
+            byte[] dataBytes = new byte[4 * 16384];
+
+            // For a full array this will build 2^14 nested indefinite length values.
+            // PeekContentBytes should return dataBytes.Slice(2, dataBytes.Length - 4)
+            //
+            // For what it's worth, the initial algorithm succeeded at 1650, and StackOverflowed with 1651.
+            //
+            // With the counter-and-no-recursion algorithm a nesting depth of 534773759 was verified,
```

Comment:
Execution time and memory usage are also possible DoS vectors. Are there any cases that deserve more scrutiny for those? In particular, are there any non-linear algorithms or cases where we'll allocate based on input without first confirming the input is that large?

Reply:
In this case the memory usage was "how big the byte[] that got passed in was".  Except for the `string` allocating members nothing allocates potentially large values, and the "most allocating" thing is the new `Stack<T>` approach to `TryCopyBitStringBytes`.  The mitigation is using the copy-to-a-span methods, or pre-guessing the encoded size by asking how big the contents span is (`PeekContentBytes()`.Length).  But there's no compression, so aside from sizeof(byte) vs sizeof(char) differences nothing exceeds the length of the user-supplied input data.

For runtime, there are a couple "2N" algorithms (count + do), but I don't think there's anything of higher order.